src/deluge/model/mod_controllable/mod_controllable_audio.cpp

```diff
@@ -1646,12 +1702,25 @@ void ModControllableAudio::endStutter(ParamManagerForTimeline* paramManager) {
 
 		UnpatchedParamSet* unpatchedParams = paramManager->getUnpatchedParamSet();
 
-		// Normally we shouldn't call this directly, but it's ok because automation isn't allowed for stutter anyway
-		if (unpatchedParams->getValue(Param::Unpatched::STUTTER_RATE) < 0) {
-			unpatchedParams->params[Param::Unpatched::STUTTER_RATE].setCurrentValueBasicForSetup(0);
+		if (runtimeFeatureSettings.get(RuntimeFeatureSettingType::QuantizedStutterRate)
+		    == RuntimeFeatureStateToggle::On) {
+			// Quantized Stutter FX (set back the value it had just before stuttering so orange LEDs are redrawn)
+			unpatchedParams->params[Param::Unpatched::STUTTER_RATE].setCurrentValueBasicForSetup(
+			    stutterer.valueBeforeStuttering);
 			view.notifyParamAutomationOccurred(paramManager);
 		}
+		else {
+			// Regular Stutter FX (if below middle value, reset it back to middle)
+			// Normally we shouldn't call this directly, but it's ok because automation isn't allowed for stutter anyway
+			if (unpatchedParams->getValue(Param::Unpatched::STUTTER_RATE) < 0) {
+				unpatchedParams->params[Param::Unpatched::STUTTER_RATE].setCurrentValueBasicForSetup(0);
+				view.notifyParamAutomationOccurred(paramManager);
+			}
+		}
 	}
+	// Reset temporary and diff values for Quantized stutter
+	stutterer.lastQuantizedKnobDiff = 0;
```

Comment:
@m-m-adams  I think having all the Quantized Stutter code inside the stutter functions (beginStutter, getStutterRate and endStutter) is better for separation of concerns. It is the stutter internal functions the ones that contain all the logic related to stutter quantization. I don't think moving the code outside to modEncoderAction is cleaner. 

The only code that is outside the stutter functions is the one controlling the knob indicator levels (making it quantized only for stutter, in a centralized place so all stutters behave the same, that is, global song, single clip, kit affect all, etc), which is a UI thing, so it is fine that it is outside.

The tricky thing here with this quantized-stutter feature is that the same parameter is controlling a two-states thing like stutter which has a different range while not pressed, but another range while stuttering (that is, when stutter knob pressed, the value and indicator are then reset to zero (center position) to allow you to change rate in both directions while allowing you to come back to the center. So that's why we need to store in the Stutterer struct the reference rate value that we capture when pressing the knob, and the relative value that is when you move the knob while pressed away from the "center".

On the other hand, the comment you made on Discord regarding value-knob calculations make sense, because there are functions already calculating that, but they are not a global static function, i need a ParamCollection object. How can i get access to that?

Reply:
You'd get access to it by doing the calculation where the stutter values are set.

I think that makes the code much cleaner as well. If the actual param manager value is accurate then there's no need to change the behaviour when turning the knob while stuttering, it will already be centered around the correct value

Reply:
There's already stutter specific code at line 854 of view.cpp to block automation - you could extend that